design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,480 @@
+  * `Failed` if the VMI migration has failed or there are rejected volumes
+
+## Design details and motivation
+
+### Motivation on the approaches
+
+There a multiple ways to perform the storage migration using libvirt. It is possible to simply copy the storage to the new destination and then swap the original disk with the new one. This approach is the most efficient as the overhead is caused only by the storage copy.
+
+A different approach is to simultaneously move the storage and live-migrate the guest. When the destination storage is not available on the host where the VM is executing, this solution is required. For this case, costs are greater since there is storage as well as live VM migration overhead.
+
+As KubeVirt is designed today and due to Kubernetes design principles of volume pod immutability, it isn't possible to hotplug and unplug new volumes while the pod is still running. KubeVirt already supports volume hotplug by attaching a new volume to a dummy pod and then bind mounting the storage in the `virt-launcher` pod. Sadly, this method fails to address the issue of removing the old PVCs that were expressly declared in the virt-launcher pod.
+Declaring all volumes hotpluggable could be a potential remedy, but we prefer to avoid it because it has been shown to be fragile for some storage providers and opaque to Kubernetes.
+
+The only workable and generic solution for this restriction appears to be to live migrate the VM with storage. Live migration involves the creation of a new virt-launcher pod by KubeVirt that has the destination PVCs attached.
+
+This migrating to a new storage provider is an exceptional operation, and very resource intensive. If in the future, the k8s community will consider the volume mutability in the pod spec, then the lighter approach could be taken into consideration.
+
+### Design
+
+This feature adds an additional controller for the CRD `Volume Migration` to virt-controller. Once the VolumeMigration object is created then KubeVirt will update the VMI status with the source and destination volume pairs.
+
+Afterwards, the controller will create a VirtualMachineInstanceMigration to trigger the VM live migration.
+
+If there are some migrated volumes in the VMI status, the VM migration controller will replace the destination volumes by creating the target virt-launcher pod with the new volumes and virt-launcher will select the source volumes to be migrated.
+
+Once the migration completes, the controller will update the migrated volumes first in the VMI spec. This will trigger the update of the volumes in the VM spec if it exists.
+
+DataVolumes are taken into account in the current proposal. DataVolumes are used to import the image during VM provisioning, but they may essentially be thought of as PVCs for the rest of the VM life cycle. According to this proposal, KubeVirt will replace the source DataVolumes with the destination PVC when a user wishes to migrate data to a PVC.
```

Comment:
Given the example below it looks like dataVolumeTemplates are removed from the VM definition.  May want to add that here?  Also not sure that is the "right" thing to do.

Reply:
Yes, Datavolumes are removed from the VM volumes and we replace the source claim with the destination. 
Keeping datavolumes doesn't make much sense to me. 
We are migrating to another PVC (with another claim name). If we need to keep DVs around then we need a new mechanism to update datavolumes to the new claim name and the assumption that the dv name = pvc name doesn't work anymore. IMO, this makes the entire design and implementation more complicated.

Or do you see any reasons for not replacing DVs?

Reply:
I totally get not bothering with DataVolumes when they are stand alone.  I'm just not as sure in the dataVolumeTemplate case.

FYI:  for DR I will be adding code to virt-controller to handle creating DVs when PVCs are created out of band

Reply:
You are right, here, I was a bit naive. I thought I could leave untouched the dataVolumeTemplate and only update the vm.Spec.Template.Spec.Volumes. However, I have just tried when the source pvc is deleted then the datavolume import is triggered again :/.
 Maybe with the Retain policy this won't happen since the original PVC still exist. Do you think I can delete also the  source DV together with the source PVC? I'd like to avoid to mess up with the owner reference of the DV.
But definitely this part isn't working properly

Reply:
I just checked, with the `Retain` policy, the DV import isn't re-triggered.

Reply:
@mhenriks this commit https://github.com/kubevirt/kubevirt/pull/10536/commits/59c888d65d195302f6a20c6a128138efb8273d64 describes a possible solution.

If the reclaim policy is delete, then we delete the corresponding DVs for the source PVC.

In all the cases, if there is A DV associated to one of the source migrated volume, then we remove it from the datavolumeTemplate during the VM spec update. Basically, after a successful volume migration. Does it sound reasonable to you?

Reply:
As we discussed offline, we decided to remove the DVs from the VM specification. In the case, of datavolumes templates, also this field is remove from the VM spec. I added more details in the proposal